### PR TITLE
Sed slide

### PR DESCRIPTION
**Pull request by mc4117**
Adding a sediment slide mechanism to the exner equation. This is used to dampen slopes and ensure unphysical slopes do not form during the simulation.

### CODE REVIEW

File: thetis/sediment_model.py [hunk 1]
```diff
@@ -303,6 +309,48 @@ def get_bedload_term(self, bathymetry):
...
+        # calculate normal to the bed
+        nz = 1/sqrt(1 + (dzdx**2 + dzdy**2))
+
+        self.betaangle = asin(sqrt(1 - (nz**2)))
+        self.tanbeta = sqrt(1 - (nz**2))/nz
+
+        # calculating magnitude of added component
+        qaval = conditional(self.tanbeta - self.tanphi > 0, (1-self.options.sediment_model_options.porosity)
+                            * 0.5*(L**2)*(self.tanbeta - self.tanphi)/(cos(self.betaangle*self.options.timestep
+                                                                           * self.options.sediment_model_options.morphological_acceleration_factor)), 0)
```
Comment (tkarna) [inline thetis/sediment_model.py:346]: long line

**Review by tkarna — COMMENTED** [review]
Thanks for this update! Only minor comments

File: test/sediment/test_sed_slide.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+swe_bnd = {}
+
+uv_vector = as_vector((0.46, 0.0))
+
+swe_bnd[left_bnd_id] = {'uv': uv_vector}
+
+swe_bnd[right_bnd_id] = {'elev': Constant(4)}
+
+solver_obj.bnd_functions['shallow_water'] = swe_bnd
```
[the remark this answers is not in the document]
Comment (mc4117) [reply test/sediment/test_sed_slide.py:72]: thanks I've removed them

File: thetis/exner_eq.py [hunk 3]
```diff
@@ -128,6 +131,24 @@ def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=No
...
+    Term which adds component to bedload transport to ensure the slope angle does not exceed a certain value
+    """
+    def residual(self, solution, solution_old, fields, fields_old, bnd_conditions=None):
+        f = 0
+
+        diff_tensor = self.sediment_model.get_sediment_slide_term(solution)
+
+        diff_flux = dot(diff_tensor, grad(-solution))
+        f += inner(grad(self.test), diff_flux)*dx
+        f += -avg(self.sediment_model.sigma)*inner(jump(self.test, self.sediment_model.n), dot(avg(diff_tensor), jump(solution, self.sediment_model.n)))*dS
```
[the remark this answers is not in the document]
Comment (mc4117) [reply thetis/exner_eq.py:145]: fixed

File: thetis/sediment_model.py (hunk 1 above)
Comment (mc4117) [reply thetis/sediment_model.py:346]: fixed

File: thetis/options.py [hunk 4]
```diff
@@ -521,10 +521,12 @@ class SedimentModelOptions(FrozenHasTraits):
     solve_suspended_sediment = Bool(False, help='Solve suspended sediment transport equation').tag(config=True)
     use_sediment_conservative_form = Bool(False, help='Solve 2D sediment transport in the conservative form').tag(config=True)
     use_bedload = Bool(False, help='Use bedload transport in sediment model').tag(config=True)
+    use_sediment_slide = Bool(False, help='Use sediment slide mechanism in sediment model').tag(config=True)
     use_angle_correction = Bool(True, help='Switch to use slope effect angle correction').tag(config=True)
     use_slope_mag_correction = Bool(True, help='Switch to use slope effect magnitude correction').tag(config=True)
     use_secondary_current = Bool(False, help='Switch to use secondary current for helical flow effect').tag(config=True)
     average_sediment_size = FiredrakeScalarExpression(None, allow_none=True, help='Average sediment size').tag(config=True)
+    slide_region = NonNegativeFloat(None, allow_none=True, help='Region where sediment slide occurs. If None then sediment slide occurs over whole domain.').tag(config=True)
```
[the remark this answers is not in the document]
Comment (mc4117) [reply thetis/options.py:529]: Good idea! Thanks for the suggestion

File: thetis/options.py [hunk 5]
```diff
@@ -534,6 +536,10 @@ class SedimentModelOptions(FrozenHasTraits):
         """).tag(config=True)
     porosity = FiredrakeCoefficient(
         Constant(0.4), help="Bed porosity for exner equation").tag(config=True)
+    max_angle = FiredrakeConstantTraitlet(
+        Constant(32), help="Angle of repose for sediment slide mechanism in degrees").tag(config=True)
+    meshgrid_size = FiredrakeConstantTraitlet(
+        Constant(0), help="Average meshgrid size for sediment slide mechanism")
```
[the remark this answers is not in the document]
Comment (mc4117) [reply thetis/options.py:542]: I have made the help message clearer and renamed the variable to sed_slide_length_scale. In general this should be the mean element size as you say but there is no real hard and fast rule so I'd prefer that the user have the option to set the parameter to a value of their choosing.

**Review by tkarna — DISMISSED** [review]
Thanks for the updates, looks good!

**Review by jwallwork23 — APPROVED** [review]
Approving rebase on top of master

**Pull request merged**